src/Controller/RecoveryController.php

```diff
@@ -0,0 +1,300 @@
+                        return $this->render(
+                            'Recovery/reset_password.html.twig',
+                            [
+                                'form' => $recoveryResetPasswordForm->createView(),
+                            ]
+                        );
+                    }
+                } else {
+                    // Verification of $email + $recoveryToken failed, start over
+                    $request->getSession()->getFlashBag()->add('error', 'flashes.recovery-reauthenticate');
+                }
+            }
+        }
+
+        return $this->redirect($this->generateUrl('recovery'));
+    }
+
+    /**
+     * @param Request $request
+     *
+     * @return Response
+     *
+     * @throws \Exception
+     */
+    public function recoveryTokenAction(Request $request): Response
+    {
+        $user = $this->getUser();
+
+        $recoveryTokenObject = new RecoveryToken();
+        $form = $this->createForm(RecoveryTokenType::class, $recoveryTokenObject);
+
+        if ('POST' === $request->getMethod()) {
+            $form->handleRequest($request);
+
+            if ($form->isSubmitted() && $form->isValid()) {
+                $user->setPlainPassword($recoveryTokenObject->password);
+
+                $this->recoveryTokenHandler->create($user);
+                $recoveryToken = $user->getPlainRecoveryToken();
+                $user->eraseToken();
+
+                $recoveryTokenAck = new RecoveryTokenAck();
+                $recoveryTokenAck->setRecoveryToken($recoveryToken);
+                $recoveryTokenAckForm = $this->createForm(
+                    RecoveryTokenAckType::class,
+                    $recoveryTokenAck,
+                    [
+                        'action' => $this->generateUrl('user_recovery_token_ack'),
+                        'method' => 'post',
+                    ]
+                );
+
+                return $this->render('User/recovery_token.html.twig',
+                    [
+                        'form' => $recoveryTokenAckForm->createView(),
+                        'recovery_token' => $recoveryToken,
+                        'recovery_secret_set' => $user->hasRecoverySecret(),
+                    ]
+                );
+            }
+        }
+
+        $recoveryTokenAckForm = $this->createForm(
+            RecoveryTokenAckType::class,
+            new RecoveryTokenAck(),
+            [
+                'action' => $this->generateUrl('user_recovery_token_ack'),
+                'method' => 'post',
+            ]
+        );
+
+        return $this->render('User/recovery_token.html.twig',
+            [
+                'form' => $form->createView(),
+                'recovery_secret_set' => $user->hasRecoverySecret(),
+            ]
+        );
+    }
+
+    /**
+     * @param Request $request
+     *
+     * @return Response
+     */
+    public function recoveryTokenAckAction(Request $request): Response
```

Comment:
This is related to the registration process, right? Then we should move this action to the RegistrationController.

Reply:
Nope, it's the form that checks whether the token has been acknowledged. This happens *both* in registration process and in user settings -> recovery token. That's why there's `registerRecoveryTokenAckAction` in RegistrationController and `recoveryTokenAckAction` in RecoveryController.